### PR TITLE
[WIP] M3-3964 POC prefetch Domains on hover

### PR DESCRIPTION
**Pull request by johnwcallahan**
## Description

This is a POC of a new pattern where we prefetch Redux data on hover. In this PR, Domains are requested when you hover over the "Domains" link in the Primary Nav. 

This only happens if we _haven't_ requested Domains for Redux before. I.e. there's no "refresh interval" like in `useReduxLoad()`. If there's already Domains data in the store, it's up to the consumer of the data to refresh it when appropriate.

If we like this pattern it can be extended to other entity types in the Primary Nav. I think it would also be beneficial to eventually do this with non-Redux pieces of data like Profile Tabs and Linode Detail Tabs. This is trickier since that data isn't currently in Redux, but we can cross that bridge if/when we get there.

## Type of Change
- Non breaking change ('update', 'change')

## Note to Reviewers

To really see the benefit of this, use the PR build and compare with production on an account with a lot of domains. Load the Dashboard, then click on the "Domains" link in the Primary Nav.

Please test various sections of the app, following the network tab to make sure that the request is made only when expected (when there's no Domains Redux data).

### CODE REVIEW

File: packages/manager/src/hooks/usePreFetch.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * const preFetch = usePreFetch();
+ *
+ * <div onmouseenter={() => preFetch(['linodes', 'images'])}>Linodes</div>
+ */
+export const usePreFetch = () => {
+  const { domains, requestDomains } = useDomains();
+
+  return (entitiesToFetch: PreFetchEntity[]) => {
+    entitiesToFetch.forEach(thisPreFetchEntity => {
+      if (
```
Comment (johnwcallahan) [inline packages/manager/src/hooks/usePreFetch.ts:26]: I know this is kind of verbose and might not scale well when other types of entities are supported, but I'd rather start out this way and make it more dynamic later if necessary.

File: packages/manager/src/hooks/usePreFetch.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        thisPreFetchEntity === 'domains' &&
```
Comment (Jskobos) [inline packages/manager/src/hooks/usePreFetch.ts:27]: I'm guessing this is meant to be a switch but the linter won't let you do that without enough `case`s

File: packages/manager/src/hooks/usePreFetch.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import useDomains from 'src/hooks/useDomains';
```
Comment (Jskobos) [inline packages/manager/src/hooks/usePreFetch.ts:1]: This is essentially the same as the `requestDeps` helper in `useReduxLoad`: going through a list of entity names and determining whether to request them or not. We probably ought to use a shared function for this. Your way is cleaner (we didn't have the other hooks back then), we can extract and reuse it once it's complete.

**Comment by johnwcallahan** [comment]
@Jskobos great feedback. I reworked this code to address the issues you raised.

The hook now accepts a requestFn param and condition (inversion of control) which I think will be better in the long run when we want to do this with non-redux things.

File: packages/manager/src/components/PrimaryNav/PrimaryNav.tsx [hunk 4]
```diff
@@ -202,16 +216,30 @@ export const PrimaryNav: React.FC<Props> = props => {
...
+          const props = {
+            key: thisLink.display,
+            closeMenu,
+            isCollapsed,
+            locationSearch: location.search,
+            locationPathname: location.pathname,
+            ...thisLink
+          };
+
+          // PrefetchPrimaryLink and PrimaryLink are two separate components because invocation of
```
Comment (Jskobos) [inline packages/manager/src/components/PrimaryNav/PrimaryNav.tsx:229]: This is good, or call `usePrefetch` with an empty array?

File: packages/manager/src/hooks/usePreFetch.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import * as React from 'react';
+
+export const PREFETCH_DELAY = 150;
```
Comment (Jskobos) [inline packages/manager/src/hooks/usePreFetch.ts:3]: This delay works great, only happens when I'm about to click somewhere.

File: packages/manager/src/components/PrimaryNav/PrimaryNav.tsx [hunk 6]
```diff
@@ -372,3 +407,25 @@ const PrimaryLink: React.FC<PrimaryLinkProps> = React.memo(props => {
...
+export const PrefetchPrimaryLink: React.FC<PrimaryLinkProps &
+  PrefetchPrimaryLinkProps> = React.memo(props => {
+  const { makeRequest, cancelRequest } = usePrefetch(
+    props.prefetchRequestFn,
+    props.prefetchRequestCondition
+  );
+
+  const prefetchProps: PrimaryLinkProps['prefetchProps'] = {
+    onMouseEnter: makeRequest,
+    onFocus: makeRequest,
```
Comment (Jskobos) [inline packages/manager/src/components/PrimaryNav/PrimaryNav.tsx:426]: Did you debate `onBlur: cancelRequest` too? I think it's fine without it but worth considering

Comment (johnwcallahan) [reply packages/manager/src/components/PrimaryNav/PrimaryNav.tsx:427]: Oh yeah, good catch. Just added this.

**Pull request merged**